### PR TITLE
NMS-10191: The KSC Dashlet for the Ops-Board is not working

### PR DESCRIPTION
**Pull request by patrick-schweizer**
* JIRA: https://issues.opennms.org/browse/NMS-10191

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 1]
```diff
@@ -430,6 +425,19 @@ public Component getComponent() {
         });
     }
 
+    String determineNodeIdByResourceId(String resourceId){
+
+        // old implementation:
+        String[] arr0 = resourceId.split("\\.");
```
Comment (patrick-schweizer) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: I am not sure under which circumstances this code has worked previously but it doesn't seem to work anymore.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 2]
```diff
@@ -430,6 +425,19 @@ public Component getComponent() {
...
+        String[] arr1 = arr0[0].split("[\\[\\]]");
+        // data.put("nodeId", arr1[1]); // results in: Test:1525957453778 which is not a valid node id
+
+        // possible new solution but with ugly casting:
+        int onmsNodeId = ((OnmsNode)m_resourceDao.getResourceById(ResourceId.fromString(resourceId))
+                .getParent().getEntity()).getId();
```
Comment (patrick-schweizer) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: is there a better way to deduct the internal node id from a resourceId? The above code works but looks mighty ugly to me...

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 3]
```diff
@@ -405,23 +407,16 @@ public Component getComponent() {
...
-                data.put("nodeLabel", m_nodeDao.getLabelForId(Integer.valueOf(data.get("nodeId"))));
+                String nodeIdentifier = Optional.ofNullable(nodeId).orElse(determineNodeIdByResourceId(resourceId));
+                data.put("nodeId", nodeIdentifier);
+                data.put("nodeLabel", m_nodeDao.getLabelForId(Integer.valueOf(nodeIdentifier)));
 
-                List<OnmsResource> resourceList = m_resourceDao.getResourceById(ResourceId.get("node", data.get("nodeId"))).getChildResources();
+                List<OnmsResource> resourceList = m_resourceDao.getResourceById(ResourceId.get("node", nodeIdentifier)).getChildResources();
 
                 for (OnmsResource onmsResource : resourceList) {
                     if (resourceId.equals(onmsResource.getId())) {
```
Comment (patrick-schweizer) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: I also found this problem which is unrelated to the NMS-10191:
This should never be true (and therefor we should either fix or remove this piece of code) since we compare her 2 different Types: String and ResourceId

Comment (christianpape) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: Exactly, I think this is broken since NMS-7548 where the ResourceId class was introduced. Before the resourceId was a simple String instance. Sure, this must be fixed.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java (hunk 2 above)
Comment (christianpape) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: There are some helper methods in ResourceTypeUtils and ResourceIdParser. Don't know If they can help at this point.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java (hunk 1 above)
Comment (christianpape) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: When this dashlet was originally written the nodeId and not foureignSource:foreignId was reported inside the square brackets. This behavior has changed and did break these lines of code.

Comment (patrick-schweizer) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: Thank you @christianpape, I'll change the code

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java (hunk 2 above)
Comment (patrick-schweizer) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: +1 I'll take a look.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java (hunk 3 above)
Comment (patrick-schweizer) [reply features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: ok Ill fix it as well. Thanks!

File: features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        when(resource.getParent()).thenReturn(parentResource);
+        when(parentResource.getEntity()).thenReturn(node);
+
+        KscDashlet dashlet = new KscDashlet(
+                this.getClass().getSimpleName(),
+                null,
+                null,
+                resourceDao,
+                null);
+        assertEquals(Integer.toString(3), dashlet.determineNodeIdByResourceId(resourceIdString));
```
Comment (mvrueden) [inline features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java]: Can we just use `"3"` here instead of `Integer.toString(3)`?

**Comment by mvrueden** [comment]
Marked as `triage` as creator requested to not merge yet.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 5]
```diff
@@ -395,33 +398,23 @@ public Component getComponent() {
...
-                data.put("nodeLabel", m_nodeDao.getLabelForId(Integer.valueOf(data.get("nodeId"))));
-
-                List<OnmsResource> resourceList = m_resourceDao.getResourceById(ResourceId.get("node", data.get("nodeId"))).getChildResources();
+                Map<String, String> data = new HashMap<>();
+                ResourceId resourceId = ResourceId.fromString(resourceIdString);
+                String nodeIdentifier = Optional.ofNullable(nodeId).orElse(determineNodeIdByResourceId(resourceId));
+                data.put("nodeId", nodeIdentifier);
+                data.put("nodeLabel", m_nodeDao.getLabelForId(Integer.valueOf(nodeIdentifier)));
 
+                List<OnmsResource> resourceList = m_resourceDao.getResourceById(ResourceId.get("node", nodeIdentifier)).getChildResources();
```
Comment (j-white) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: I'd opt for calling `ResourceDao#getResourceForNode` here instead.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 6]
```diff
@@ -430,6 +423,13 @@ public Component getComponent() {
         });
     }
 
+    String determineNodeIdByResourceId(ResourceId resourceId){
+        OnmsResource parentResource = m_resourceDao.getResourceById(resourceId).getParent();
```
Comment (j-white) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: The parent can be null, in which case the we should attempt to retrieve the node directly from the given resource.

File: features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java [hunk 7]
```diff
@@ -430,6 +423,13 @@ public Component getComponent() {
...
+        OnmsNode node = ResourceTypeUtils.getNodeFromResource(parentResource);
```
Comment (j-white) [inline features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java]: We should re-use this node object instead of making subsequent NodeDao calls with the extract ID.

File: features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java (hunk 4 above)
Comment (patrick-schweizer) [reply features/vaadin-dashlets/dashlet-ksc/src/test/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashletTest.java]: well the method returns a String therefore we need to compare String with String, let me see if I can change the return parameter to int which would be better anyway.

**Comment by patrick-schweizer** [comment]
@christianpape, @j-white, @mvrueden, thank you for your input. I tried to address them all. Please take a look again. If ok it can be merged. There is also another PR for foundation-2016:
https://github.com/OpenNMS/opennms/pull/2023
